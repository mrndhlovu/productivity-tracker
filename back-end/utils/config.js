const dotenv = require("dotenv");

dotenv.config();

const isDevelopment = process.env.DEVELOPMENT === "development";

const S_GRID_API_KEY = process.env.SEND_GRID_API_KEY;
const CONNECTION_URI = process.env.MONGODB_URI;
const { LOCAL_MONGO_DB } = process.env;

const PORT = process.env.PORT || 5000;

const { PRIVATE_SIGNATURE, PUBLIC_SIGNATURE } = process.env;
const BASE_URL = isDevelopment
  ? "http://localhost:3000"
  : `http://ec2-3-248-208-68.eu-west-1.compute.amazonaws.com:5000`;

const ALLOWED_UPDATE_FIELDS_USER = [
  "firstName",
  "lastName",
  "email",
  "password",
];
const ALLOWED_UPDATE_TASK_FIELDS = ["title", "dueDate", "status", "archived"];
const ALLOWED_UPDATE_LIST_FIELDS = ["title"];

module.exports = {
  ALLOWED_UPDATE_FIELDS_USER,
  ALLOWED_UPDATE_LIST_FIELDS,
  ALLOWED_UPDATE_TASK_FIELDS,
  CONNECTION_URI,
  isDevelopment,
  LOCAL_MONGO_DB,
  PORT,
  PRIVATE_SIGNATURE,
  PUBLIC_SIGNATURE,
  BASE_URL,
  S_GRID_API_KEY,
};
