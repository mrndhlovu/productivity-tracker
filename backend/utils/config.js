const dotenv = require("dotenv");

dotenv.config();

const PORT = process.env.PORT || 5000;

const S_GRID_API_KEY = process.env.SEND_GRID_API_KEY;
const CONNECTION_URI = process.env.MONGODB_URI;

const { PRIVATE_SIGNATURE, PUBLIC_SIGNATURE, LOCAL_MONGO_DB } = process.env;

const ALLOWED_UPDATE_FIELDS_USER = [
  "firstName",
  "lastName",
  "email",
  "password",
];
const ALLOWED_UPDATE_TASK_FIELDS = ["title", "dueDate", "status", "archived"];
const ALLOWED_UPDATE_LIST_FIELDS = ["title"];

module.exports = {
  ALLOWED_UPDATE_FIELDS_USER,
  ALLOWED_UPDATE_LIST_FIELDS,
  ALLOWED_UPDATE_TASK_FIELDS,
  CONNECTION_URI,
  LOCAL_MONGO_DB,
  PORT,
  PRIVATE_SIGNATURE,
  PUBLIC_SIGNATURE,
  S_GRID_API_KEY,
};
