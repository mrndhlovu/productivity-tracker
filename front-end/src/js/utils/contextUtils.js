import { createContext } from "react";

export const MainContext = createContext(null);

export const TaskContext = createContext(null);
