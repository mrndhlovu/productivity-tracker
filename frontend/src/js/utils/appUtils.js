/* eslint-disable curly */
/* eslint-disable indent */
/* eslint-disable nonblock-statement-body-position */
import moment from "moment";

export const getValidUpdateFields = (data, allowedUpdates) => {
  const filtered = Object.keys(data)
    .filter((key) => allowedUpdates.includes(key))
    .reduce((obj, key) => {
      // eslint-disable-next-line no-param-reassign
      obj[key] = data[key];
      return obj;
    }, {});

  return filtered;
};

export const resetForm = (data) => {
  if (typeof data === "object")
    return data.forEach((id) => {
      document.getElementById(id).value = "";
    });
  // eslint-disable-next-line no-return-assign
  return (document.getElementById(data).value = "");
};

export const capitalize = (string) => {
  const result = string
    ? `${string.charAt(0).toUpperCase()}${string.slice(1)}`
    : "";
  return result;
};

export const getLocaleString = () => {
  // eslint-disable-next-line newline-per-chained-call
  const result = moment().format().split(":").slice(0, 2).join(":");
  return result;
};

export const getFormattedDateString = (date, status) => {
  const taskDate = new Date(date).getTime();
  const isExpired = Date.now() > taskDate;

  // eslint-disable-next-line curly
  if (isExpired && status !== "complete ")
    return moment(taskDate).subtract(3, "days").calendar();

  return moment(taskDate).format("llll");
};

export const taskStylingClassName = (status, date) => {
  const taskDate = new Date(date).getTime();
  const isExpired = Date.now() > taskDate;

  return `${
    // eslint-disable-next-line no-nested-ternary
    (status === "incomplete" || status === "todo") && isExpired
      ? "task__expired"
      : status === "complete"
      ? "task__completed"
      : "task__todo"
  }`;
};
