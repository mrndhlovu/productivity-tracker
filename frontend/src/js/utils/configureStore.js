import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const createStoreWithMiddleware = composeEnhancers(applyMiddleware(thunk))(
  createStore,
);

export default function configureStore(initialState = {}) {
  return createStoreWithMiddleware(() => {}, initialState);
}
