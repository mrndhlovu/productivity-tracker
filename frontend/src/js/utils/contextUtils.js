import { createContext } from "react";

export const MainContext = createContext(null);

export const AlertContext = createContext(null);
