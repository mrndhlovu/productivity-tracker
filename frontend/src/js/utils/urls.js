export const baseURL =
  process.env.NODE_ENV === "development"
    ? "http://127.0.0.1:5000"
    : "http://34.245.58.144";

export const TASKS_EP = "/v1/api/tasks";
export const AUTH_EP = "/v1/api/auth";

export const PARAMS = {
  baseURL,
  headers: { Accept: "application/json", "Content-Type": "application/json" },
  credentials: "same-origin",
  withCredentials: true,
};
