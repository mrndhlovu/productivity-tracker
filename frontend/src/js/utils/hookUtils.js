import { useContext } from "react";
import { MainContext } from "./contextUtils";

// eslint-disable-next-line import/prefer-default-export
export const useMainContent = () => useContext(MainContext);
