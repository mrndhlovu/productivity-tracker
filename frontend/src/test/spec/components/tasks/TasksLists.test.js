/* eslint-disable no-return-assign */
/* eslint-disable react/jsx-props-no-spreading */

import React from "react";
import { render, fireEvent } from "@testing-library/react";
import historyData from "react-router";
import { act } from "react-dom/test-utils";
import toJSON from "enzyme-to-json";

import { MainContext } from "../../../../js/utils/contextUtils";
import { MOCK_HISTORY_PROPS, MOCK_TASKS_LIST } from "../../../utils";
import TasksListPage from "../../../../js/components/tasks/TasksList";

const CONTEXT_PROPS = {
  taskActionHandler: jest.fn(),
};

const DEFAULT_PROPS = {
  header: "To Do",
  tasks: MOCK_TASKS_LIST,
};

describe("TasksListPage", () => {
  let wrapper;

  const init = async () => {
    jest.spyOn(historyData, "useHistory").mockReturnValue(MOCK_HISTORY_PROPS);
  };

  const setup = async () => {
    await act(
      async () =>
        (wrapper = await render(
          <MainContext.Provider value={CONTEXT_PROPS}>
            <TasksListPage {...DEFAULT_PROPS} />
          </MainContext.Provider>,
        )),
    );
  };

  beforeEach(async () => {
    jest.resetModules();
    init();
    await setup();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should render task items and click the edit buttons.", async () => {
    const { getByTestId, getByText } = wrapper;

    getByText(/To Do/);
    getByTestId(/to-do-tasks-header/);
    getByTestId(/to-do-task-item-0/);

    const TASK_TITLE = MOCK_TASKS_LIST[0].title
      .split(" ")
      .join("-")
      .toLowerCase();

    const completeButton = getByTestId(`${TASK_TITLE}-check-edit-button`);
    fireEvent.click(completeButton);
    expect(CONTEXT_PROPS.taskActionHandler).toHaveBeenCalledWith(
      "complete",
      MOCK_TASKS_LIST[0],
    );

    const incompleteButton = getByTestId(`${TASK_TITLE}-x-edit-button`);
    fireEvent.click(incompleteButton);
    expect(CONTEXT_PROPS.taskActionHandler).toHaveBeenCalledWith(
      "incomplete",
      MOCK_TASKS_LIST[0],
    );

    const editButton = getByTestId(`${TASK_TITLE}-pen-edit-button`);
    fireEvent.click(editButton);
    expect(CONTEXT_PROPS.taskActionHandler).toHaveBeenCalledWith(
      "edit",
      MOCK_TASKS_LIST[0],
    );

    const deleteButton = getByTestId(`${TASK_TITLE}-delete-edit-button`);
    fireEvent.click(deleteButton);
    expect(CONTEXT_PROPS.taskActionHandler).toHaveBeenCalledWith(
      "delete",
      MOCK_TASKS_LIST[0],
    );

    expect(CONTEXT_PROPS.taskActionHandler).toHaveBeenCalledTimes(4);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
