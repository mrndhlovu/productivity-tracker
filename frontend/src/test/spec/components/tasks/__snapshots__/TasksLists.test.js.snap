// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksListPage should render task items and click the edit buttons. 1`] = `undefined`;
